// form-render/es/widgets/antd/map.js
import "antd/es/card/style";
import _Card from "antd/es/card";
import "antd/es/collapse/style";
import _Collapse from "antd/es/collapse";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

import React, { useEffect, useState } from 'react';
import cn from 'classnames';
import { useStore2 } from '../../form-render-core/src/hooks';
var Panel = _Collapse.Panel;
export default function Map(_ref) {
  var children = _ref.children,
      title = _ref.title,
      schema = _ref.schema;

  var _useState = useState(schema.collapsed || false),
      _useState2 = _slicedToArray(_useState, 2),
      collapsed = _useState2[0],
      setCollapsed = _useState2[1];

  var _useStore = useStore2(),
      globalTheme = _useStore.theme,
      globalDisplayType = _useStore.displayType;

  var theme = schema.theme || globalTheme;
  var displayType = schema.displayType || globalDisplayType;
  useEffect(function () {
    if (schema.hasOwnProperty('collapsed')) {
      setCollapsed(schema.collapsed);
    }
  }, [schema.collapsed]);

  if (!title) {
    return /*#__PURE__*/React.createElement("div", {
      className: "w-100"
    }, children);
  }

  if (theme === 'tile') {
    return /*#__PURE__*/React.createElement("div", {
      className: "w-100"
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        fontSize: 17,
        fontWeight: 500,
        paddingBottom: 4,
        borderBottom: '1px solid rgba( 0, 0, 0, .2 )',
        marginBottom: 16
      }
    }, title, /*#__PURE__*/React.createElement("span", {
      className: "fr-desc ml2"
    }, (schema === null || schema === void 0 ? void 0 : schema.description) ? "( ".concat(schema.description, " )") : '')), /*#__PURE__*/React.createElement("div", {
      style: {
        marginLeft: displayType == 'row' ? 0 : 12
      }
    }, children));
  } // 新增卡片视图


  if (theme === 'card') {
    return /*#__PURE__*/React.createElement(_Card, {
      id: title,
      title: /*#__PURE__*/React.createElement(React.Fragment, null, title, /*#__PURE__*/React.createElement("span", {
        className: "fr-desc ml2"
      }, (schema === null || schema === void 0 ? void 0 : schema.description) ? "( ".concat(schema.description, " )") : '')),
      className: "fr-theme-card-wrap"
    }, children);
  }

  var toggle = function toggle(keyList) {
    if (keyList.length > 0) {
      setCollapsed(false);
    } else {
      setCollapsed(true);
    }
  };

  return /*#__PURE__*/React.createElement("div", {
    className: "w-100"
  }, /*#__PURE__*/React.createElement(_Collapse, {
    activeKey: collapsed ? [] : ['1'],
    onChange: toggle,
    bordered: theme !== 'collapse:pure',
    ghost: theme === 'collapse:ghost'
  }, /*#__PURE__*/React.createElement(Panel, {
    header: /*#__PURE__*/React.createElement("span", {
      style: {
        fontSize: 16,
        fontWeight: 500
      }
    }, title, /*#__PURE__*/React.createElement("span", {
      className: "fr-desc ml2"
    }, (schema === null || schema === void 0 ? void 0 : schema.description) ? "( ".concat(schema.description, " )") : '')),
    key: "1",
    className: "fr-collapse-object"
  }, /*#__PURE__*/React.createElement("div", {
    className: cn({
      'fr-collapse-object-child-row': displayType === 'row',
      'fr-collapse-object-child-column': displayType === 'column'
    })
  }, children))));
}
